unsafe-fields/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+    field: F,
+}
+
+// NOTE on design: It may seem counter-intuitive to offer an impl of traits that
+// don't require `unsafe` to call. Unfortunately, this is a fundamental
+// requirement if users want to be able to mark their types as `Copy`. Luckily,
+// we can implement `Copy` (and its unavoidable super-trait, `Clone`) without
+// invoking user code or opening up the possibility of modifying the field. We
+// do this by only implementing `Copy` and `Clone` when `F: Copy`. For `Clone`,
+// the user is still able to provide a manual impl, so this does not
+// fundamentally restrict what behavior can be supported.
+impl<O: ?Sized, F: Copy, const NAME_HASH: u128> Copy for Unsafe<O, F, { NAME_HASH }> {}
+impl<O: ?Sized, F: Copy, const NAME_HASH: u128> Clone for Unsafe<O, F, { NAME_HASH }> {
+    #[allow(clippy::non_canonical_clone_impl)]
+    fn clone(&self) -> Self {
+        // SAFETY: We don't call any user-defined code here (only make a
+        // bit-for-bit copy of `self.field`), so there's no way to accidentally
+        // invoke user-defined code or modify `self.field`.
+        Unsafe { _marker: PhantomData, field: self.field }
+    }
+}
+
+impl<O: ?Sized, F: ?Sized, const NAME_HASH: u128> Unsafe<O, F, { NAME_HASH }> {
+    /// Gets a reference to the inner value.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for upholding any safety invariants associated
+    /// with this field.
+    #[inline(always)]
+    pub const unsafe fn as_ref(&self) -> &F {
+        // SAFETY: This method is unsafe to call.
+        &self.field
+    }
+
+    /// Gets a mutable reference to the inner value.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for upholding any safety invariants associated
+    /// with this field.
+    #[inline(always)]
+    pub unsafe fn as_mut(&mut self) -> &mut F {
+        // SAFETY: This method is unsafe to call.
+        &mut self.field
+    }
+}
+
+impl<O: ?Sized, F, const NAME_HASH: u128> Unsafe<O, F, { NAME_HASH }> {
+    /// Constructs a new `Unsafe<O, F, NAME_HASH>`.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for upholding any safety invariants associated
+    /// with this field.
+    #[inline(always)]
+    pub const unsafe fn new(field: F) -> Unsafe<O, F, { NAME_HASH }> {
+        // SAFETY: This method is unsafe to call.
+        Unsafe { _marker: PhantomData, field }
+    }
+
+    /// Extracts the inner `F` from `self`.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for upholding any safety invariants associated
+    /// with this field.
+    #[inline(always)]
+    pub const unsafe fn into(self) -> F {
+        use core::mem::ManuallyDrop;
+
+        let slf = ManuallyDrop::new(self);
+
+        #[repr(C)]
+        union Transmute<O: ?Sized, F, const NAME_HASH: u128> {
+            src: ManuallyDrop<Unsafe<O, F, { NAME_HASH }>>,
+            dst: ManuallyDrop<F>,
+        }
```

Comment:
Unless there's good reason to do otherwise, I'd prefer to see this written as:
```suggestion
        #[repr(C)]
        union Transmute<Src, Dst> {
            src: ManuallyDrop<Src>,
            dst: ManuallyDrop<Dst>,
        }
```

Reply:
Done.